File: backend/src/upload/upload.controller.ts
import {
  Controller,
  Post,
  Get,
  Param,
  UploadedFile,
  UseInterceptors,
  Res,
  ParseIntPipe,
  BadRequestException,
  NotFoundException,
} from '@nestjs/common';
import { FileInterceptor } from '@nestjs/platform-express';
import { Response } from 'express';
import { UploadService } from './upload.service';
import { WordService } from '../word/word.service';

@Controller('api')
export class UploadController {
  constructor(
    private readonly uploadService: UploadService,
    private readonly wordService: WordService,
  ) {}

  @Post('upload')
  @UseInterceptors(FileInterceptor('file'))
  async uploadExcel(@UploadedFile() file: Express.Multer.File) {
    if (!file) {
      throw new BadRequestException('Файл не предоставлен');
    }
    const result = await this.uploadService.processExcel(
      file.buffer,
      file.originalname,
    );
    return result;
  }

  @Get('files')
  async getFiles() {
    return this.uploadService.getFiles();
  }

  @Get('files/:id/rows')
  async getFileRows(@Param('id', ParseIntPipe) id: number) {
    return this.uploadService.getFileRows(id);
  }

  @Get('rows/:id/word')
  async generateWord(
    @Param('id', ParseIntPipe) id: number,
    @Res() res: Response,
  ) {
    const row = await this.uploadService.getRow(id);
    if (!row) {
      throw new NotFoundException('Строка не найдена');
    }
    const buffer = await this.wordService.generatePaymentRequirement(row);

    res.set({
      'Content-Type':
        'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
      'Content-Disposition': `attachment; filename="payment_requirement_${row.id}.docx"`,
    });
    res.send(buffer);
  }
}

File: backend-kotlin/src/main/kotlin/com/excelpilot/controller/UploadController.kt
package com.excelpilot.controller

import com.excelpilot.service.UploadService
import com.excelpilot.service.WordService
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import org.springframework.web.server.ResponseStatusException

@RestController
@RequestMapping("/api")
@CrossOrigin
class UploadController(
    private val uploadService: UploadService,
    private val wordService: WordService,
) {

    @PostMapping("/upload")
    fun uploadExcel(@RequestParam("file") file: MultipartFile): ResponseEntity<Any> {
        if (file.isEmpty) {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, "No file provided")
        }
        val result = uploadService.processExcel(file.bytes, file.originalFilename ?: "unknown.xlsx")
        return ResponseEntity.ok(result)
    }

    @GetMapping("/files")
    fun getFiles(): ResponseEntity<Any> {
        return ResponseEntity.ok(uploadService.getFiles())
    }

    @GetMapping("/files/{id}/rows")
    fun getFileRows(@PathVariable id: Long): ResponseEntity<Any> {
        return ResponseEntity.ok(uploadService.getFileRows(id))
    }

    @GetMapping("/rows/{id}/word")
    fun generateWord(@PathVariable id: Long): ResponseEntity<ByteArray> {
        val row = uploadService.getRow(id)
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "Row not found")

        val docxBytes = wordService.generatePaymentRequirement(row)

        val headers = HttpHeaders()
        headers.contentType = MediaType.parseMediaType(
            "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
        )
        headers.setContentDispositionFormData("attachment", "payment_requirement_$id.docx")

        return ResponseEntity.ok().headers(headers).body(docxBytes)
    }
}
